finetune/lora.py

```diff
@@ -0,0 +1,225 @@
+"""
+Instruction-tuning with LoRA on the Alpaca dataset.
+
+Note: If you run into a CUDA error "Expected is_sm80 to be true, but got false", uncomment the line
+`torch.backends.cuda.enable_flash_sdp(False)` in the script below (see https://github.com/Lightning-AI/lit-llama/issues/101).
+"""
+import sys
+from typing import Literal
+from pathlib import Path
+import os
+import time
+
+import lightning as L
+import numpy as np
+import torch
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from generate.base import generate
+from lit_parrot.lora import mark_only_lora_as_trainable, lora, lora_state_dict
+from lit_parrot.model import Parrot, Config
+from lit_parrot.tokenizer import Tokenizer
+from lit_parrot.utils import lazy_load, check_valid_checkpoint_dir # TODO: Use lazy_load
+from scripts.prepare_alpaca import generate_prompt
+
+
+instruction_tuning = True
+eval_interval = 100
+save_interval = 100
+eval_iters = 100
+log_interval = 1
+
+# Hyperparameters
+learning_rate = 3e-4
+batch_size = 128
+micro_batch_size = 4
+gradient_accumulation_iters = batch_size // micro_batch_size
+assert gradient_accumulation_iters > 0
+max_iters = 50000 * 3 // micro_batch_size
+weight_decay = 0.0
+max_seq_length = 256  # see scripts/prepare_alpaca.py
+lora_r = 8
+lora_alpha = 16
+lora_dropout = 0.05
+warmup_iters = 100
+
+
+def main(
+    data_dir: Path = Path("data/alpaca"),
+    checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    out_dir: Path = Path("out/lora/alpaca"),
+    precision: Literal["bf16-true", "32-true", "bf16-mixed"] = "bf16-true",
+):
+
+    print(checkpoint_dir)
+    check_valid_checkpoint_dir(checkpoint_dir)
+
+    fabric = L.Fabric(accelerator="cuda", devices=1, precision=precision)
+    fabric.launch()
+    fabric.seed_everything(1337 + fabric.global_rank)
+
+    if fabric.global_rank == 0:
+        os.makedirs(out_dir, exist_ok=True)
+
+    train_data, val_data = load_datasets(data_dir=data_dir)
+
+    config = Config.from_name(name=checkpoint_dir.name, block_size=max_seq_length)
+    config.block_size = max_seq_length
+
+    checkpoint_path = checkpoint_dir / "lit_model.pth"
+   
+    fabric.print(f"Loading model {str(checkpoint_path)!r} with {config.__dict__}")
+    
+    #with fabric.init_module():
+    #    model = Parrot(config)
+    #with lazy_load(checkpoint_dir / "lit_model.pth") as checkpoint:
+    #    # strict=False because missing keys due to adapter weights not contained in state dict
+    #    model.load_state_dict(checkpoint, strict=False)
```

Comment:
What is the issue with `lazy_load`?

Reply:
@carmocca It resulted in shape issues. It seems like it couldn't load the tensors properly. I can try again an print the exact error message.